File: packages/akribes-sdk-python/akribes_sdk/resources/mcp.py
from __future__ import annotations

from urllib.parse import quote

from akribes_sdk.models import (
    McpDriftResult,
    McpHealth,
    McpRefreshResult,
    McpServerSummary,
    McpToolSummary,
)
from akribes_sdk._parsers import (
    parse_mcp_server_summary,
    parse_mcp_tool_summary,
    parse_mcp_health,
    parse_mcp_refresh_result,
    parse_mcp_drift_result,
)
from akribes_sdk.resources._base import ProjectResource


class Mcp(ProjectResource):
    """MCP server/tool discovery for a project."""

    async def list_servers(self) -> list[McpServerSummary]:
        res = await self._request("GET", self._project_url("mcp", "servers"))
        return [parse_mcp_server_summary(s) for s in res.json()]

    async def list_tools(self) -> list[McpToolSummary]:
        res = await self._request("GET", self._project_url("mcp", "tools"))
        return [parse_mcp_tool_summary(t) for t in res.json()]

    async def health(self, alias: str) -> McpHealth:
        res = await self._request(
            "GET",
            self._project_url("mcp", "servers", quote(alias, safe=""), "health"),
        )
        return parse_mcp_health(res.json())

    async def refresh(self, alias: str) -> McpRefreshResult:
        """Force a fresh ``tools/list`` against the remote MCP server and
        update the pinned schema in the DB. Returns the new tool count."""
        res = await self._request(
            "POST",
            self._project_url("mcp", "servers", quote(alias, safe=""), "refresh"),
            json={},
        )
        return parse_mcp_refresh_result(res.json())

    async def drift(self, alias: str) -> McpDriftResult:
        """Compare the pinned schema against the remote server's live
        ``tools/list`` and report added/removed tool names."""
        res = await self._request(
            "GET",
            self._project_url("mcp", "servers", quote(alias, safe=""), "drift"),
        )
        return parse_mcp_drift_result(res.json())

File: crates/akribes-sdk/src/sub/mcp.rs
use std::sync::Arc;

use crate::client::{AkribesClient, Inner};
use crate::error::Result;
use crate::models::{
    McpDriftResult, McpHealth, McpRefreshResult, McpServerSummary, McpToolSummary,
};

/// Sub-client for MCP server/tool discovery. Obtained via
/// [`crate::client::ProjectScope::mcp()`].
#[derive(Clone, Debug)]
pub struct McpClient {
    pub(crate) inner: Arc<Inner>,
    pub(crate) project_id: i64,
}

impl McpClient {
    pub(crate) fn new(inner: Arc<Inner>, project_id: i64) -> Self {
        Self { inner, project_id }
    }

    fn c(&self) -> AkribesClient {
        AkribesClient {
            inner: Arc::clone(&self.inner),
        }
    }

    fn base(&self) -> String {
        format!("{}/projects/{}/mcp", self.inner.base_url, self.project_id)
    }

    pub async fn list_servers(&self) -> Result<Vec<McpServerSummary>> {
        let url = format!("{}/servers", self.base());
        self.c().get_list(&url).await
    }

    pub async fn list_tools(&self) -> Result<Vec<McpToolSummary>> {
        let url = format!("{}/tools", self.base());
        self.c().get_list(&url).await
    }

    pub async fn health(&self, alias: &str) -> Result<McpHealth> {
        let url = format!(
            "{}/servers/{}/health",
            self.base(),
            urlencoding::encode(alias)
        );
        let res = self.c().get_opt::<McpHealth>(&url).await?;
        res.ok_or_else(|| crate::error::AkribesError::HttpStatus {
            status: 404,
            message: format!("mcp server '{}' not found", alias),
        })
    }

    /// Force a fresh `tools/list` against the remote MCP server and update
    /// the pinned schema in the DB.
    pub async fn refresh(&self, alias: &str) -> Result<McpRefreshResult> {
        let url = format!(
            "{}/servers/{}/refresh",
            self.base(),
            urlencoding::encode(alias)
        );
        self.c().post(&url, &serde_json::json!({})).await
    }

    /// Compare the pinned schema against the remote server's live `tools/list`
    /// and report added/removed tool names.
    pub async fn drift(&self, alias: &str) -> Result<McpDriftResult> {
        let url = format!(
            "{}/servers/{}/drift",
            self.base(),
            urlencoding::encode(alias)
        );
        let res = self.c().get_opt::<McpDriftResult>(&url).await?;
        res.ok_or_else(|| crate::error::AkribesError::HttpStatus {
            status: 404,
            message: format!("mcp server '{}' not found", alias),
        })
    }
}
